packages/itwin/tree-widget/src/components/trees/models-tree/internal/QueryHandler.ts

```diff
@@ -0,0 +1,457 @@
+          if (subject.parentId) {
+            pushToMap(acc.subjectsHierarchy, subject.parentId, subject.id);
+          }
+
+          if (subject.targetPartitionId) {
+            pushToMap(acc.targetPartitionSubjects, subject.targetPartitionId, subject.id);
+          }
+          return acc;
+        },
+        { subjectsHierarchy: new Map<Id64String, Set<Id64String>>(), targetPartitionSubjects: new Map<Id64String, Set<Id64String>>() },
+      ),
+      mergeMap((state) =>
+        this.runModelsQuery().pipe(
+          reduce(
+            (acc, model) => {
+              const subjectIds = acc.targetPartitionSubjects.get(model.id) ?? new Set();
+              subjectIds.add(model.parentId);
+              subjectIds.forEach((subjectId) => {
+                pushToMap(acc.subjectModels, subjectId, model.id);
+              });
+              return acc;
+            },
+            { ...state, subjectModels: new Map<Id64String, Set<Id64String>>() },
+          ),
+        ),
+      ),
+      map(({ subjectModels, subjectsHierarchy }) => ({ subjectModels, subjectsHierarchy })),
+      shareReplay(),
+    );
+  }
+
+  private runSubjectsQuery(): Observable<{ id: Id64String; parentId?: Id64String; targetPartitionId?: Id64String }> {
+    const query = /* sql */ `
+      SELECT ECInstanceId id, Parent.Id parentId, json_extract(JsonProperties, '$.Subject.Model.TargetPartition') targetPartitionId
+      FROM bis.Subject
+    `;
+    return this.runQuery(query, undefined, (row) => ({ id: row.id, parentId: row.parentId, targetPartitionId: row.targetPartitionId }));
+  }
+
+  private runModelsQuery(): Observable<{ id: Id64String; parentId: Id64String }> {
+    const query = /* sql */ `
+      SELECT p.ECInstanceId id, p.Parent.Id parentId
+      FROM bis.InformationPartitionElement p
+      INNER JOIN bis.GeometricModel3d m ON m.ModeledElement.Id = p.ECInstanceId
+      WHERE NOT m.IsPrivate
+    `;
+    return this.runQuery(query, undefined, (row) => ({ id: row.id, parentId: row.parentId }));
+  }
+
+  public queryModelCategories(id: Id64String): Observable<Id64String> {
+    return this.getObservableOrInsertToCache({
+      cache: this._modelCategoriesCache,
+      cacheKey: id,
+      factory: () => this.runModelCategoriesQuery(id),
+    });
+  }
+
+  private runModelCategoriesQuery(id: Id64String): Observable<Id64String> {
+    const bindings = new Array<Id64String>();
+    const query = /* sql */ `
+      SELECT ECInstanceId id
+      FROM bis.SpatialCategory c
+      WHERE EXISTS (
+        SELECT 1
+        FROM bis.GeometricElement3d e
+        WHERE
+          ${bind("e.Model.id", id, bindings)}
+          AND e.Category.Id = c.ECInstanceId
+          AND e.Parent IS NULL
+        LIMIT 1
+      )
+    `;
+    return this.runQuery(query, [id], (row) => row.id);
+  }
+
+  public queryElementChildren(elementId: string): Observable<Id64String> {
+    const cachedChildren = this._elementHierarchyCache.get(elementId);
```

Comment:
I understand the first step to get all children IDs (filtered using always/never draw lists).

I don't understand the second one, though. Are we running those queries when determining visibility of the parent node, or do you mean we run them when the parent node gets expanded? In case of the latter, that happens after a user action, so IMO it's okay to run separate queries to determine state of the newly loaded child nodes. 

On the other hand, let's not forget that always / never drawn lists will generally contain a relatively small set of ids. Maybe, when we get the visibility change notification, we could immediately issue a query that selects model, category and element ID of all elements in those sets? That way we would only be running the query once. And caching these ids shouldn't be a problem, based on the assumption that those two lists aren't huge.